In [1]:
# Importing Required Libraries
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score


In [24]:
# Loading the Dataset
df = pd.read_csv("credit_risk_synthetic.csv")
df.head()


,age,income,loan_amount,loan_term_months,credit_score,employment_years,num_prior_defaults,home_ownership,purpose,default
0,41,165766,575961,40,647,7,0,rent,auto,1
1,28,140095,768566,35,719,10,0,mortgage,education,1
2,46,272081,581996,52,717,0,0,rent,personal,1
3,47,250223,1171649,47,584,5,0,mortgage,personal,1
4,18,51425,928936,6,636,9,1,rent,education,1


In [26]:
# Dataset Overview
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   age                 2000 non-null   int64 
 1   income              2000 non-null   int64 
 2   loan_amount         2000 non-null   int64 
 3   loan_term_months    2000 non-null   int64 
 4   credit_score        2000 non-null   int64 
 5   employment_years    2000 non-null   int64 
 6   num_prior_defaults  2000 non-null   int64 
 7   home_ownership      2000 non-null   object
 8   purpose             2000 non-null   object
 9   default             2000 non-null   int64 
dtypes: int64(8), object(2)
memory usage: 156.4+ KB


In [27]:
# Check for missing values
print(df.isnull().sum())

age                   0
income                0
loan_amount           0
loan_term_months      0
credit_score          0
employment_years      0
num_prior_defaults    0
home_ownership        0
purpose               0
default               0
dtype: int64


In [28]:
# Encoding Categorical Features

label_encoder = LabelEncoder()

df['home_ownership'] = label_encoder.fit_transform(df['home_ownership'])
df['purpose'] = label_encoder.fit_transform(df['purpose'])


In [29]:
# Feature and Target Separation
X = df.drop('default', axis=1)
y = df['default']


In [31]:
# Detect outliers using Z-score
z_scores = np.abs((X - X.mean()) / X.std())
outliers = np.where(z_scores > 3)

# Remove outliers
X_cleaned = X[(z_scores < 3).all(axis=1)]
y_cleaned = y[(z_scores < 3).all(axis=1)]

print("Number of data points after removing outliers:", len(X_cleaned))

Number of data points after removing outliers: 1902


In [32]:
# Feature Scaling
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [33]:
# Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)


In [34]:
# Logistic Regression Model Training
log_reg = LogisticRegression(max_iter=1000)
log_reg.fit(X_train, y_train)


LogisticRegression(max_iter=1000)

In [37]:
# Logistic Regression Evaluation
y_pred_log = log_reg.predict(X_test)
log_accuracy = accuracy_score(y_test, y_pred_log)

print("Logistic Regression Accuracy:", log_accuracy)


Logistic Regression Accuracy: 0.9875


In [36]:
# Support Vector Machine Model Training
svm_model = SVC(kernel='linear')
svm_model.fit(X_train, y_train)


SVC(kernel='linear')

In [38]:
# Support Vector Machine Evaluation
y_pred_svm = svm_model.predict(X_test)
svm_accuracy = accuracy_score(y_test, y_pred_svm)

print("SVM Accuracy:", svm_accuracy)


SVM Accuracy: 0.9925


In [39]:
# Testing with a Random Data Sample
random_index = np.random.randint(0, X_test.shape[0])
sample = X_test[random_index].reshape(1, -1)

print("Actual Label:", y_test.iloc[random_index])
print("Logistic Regression Prediction:", log_reg.predict(sample)[0])
print("SVM Prediction:", svm_model.predict(sample)[0])


Actual Label: 1
Logistic Regression Prediction: 1
SVM Prediction: 1


## Classification Results

- Logistic Regression achieved **98.75%** accuracy
- Support Vector Machine achieved **99.25%** accuracy

### Summary
The comparison shows that both classifiers perform effectively. However, the Support Vector Machine produces slightly higher accuracy for this classification task.
